core/deployment/src/main/java/io/quarkus/deployment/configuration/tracker/ConfigTrackingInterceptor.java

```diff
@@ -0,0 +1,152 @@
+package io.quarkus.deployment.configuration.tracker;
+
+import static io.smallrye.config.SecretKeys.doLocked;
+
+import java.io.BufferedWriter;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.regex.Pattern;
+
+import jakarta.annotation.Priority;
+
+import org.eclipse.microprofile.config.Config;
+
+import io.quarkus.deployment.configuration.BuildTimeConfigurationReader;
+import io.quarkus.runtime.LaunchMode;
+import io.smallrye.config.ConfigSourceInterceptor;
+import io.smallrye.config.ConfigSourceInterceptorContext;
+import io.smallrye.config.ConfigValue;
+import io.smallrye.config.Priorities;
+
+/**
+ * Build configuration interceptor that records all the configuration options
+ * and their values that are read during the build.
+ */
+@Priority(Priorities.APPLICATION)
+public class ConfigTrackingInterceptor implements ConfigSourceInterceptor {
```

Comment:
Yes, it could be done better. I just didn't want to expose the interceptor itself with its public API to build steps.

Reply:
I made `ConfigTrackingInterceptor` expose `ConfigTrackingInterceptor.ReadOptionsProvider` that returns a `Map<String, String>` of the read options. And the writing part is extracted to `ConfigTrackingWriter`.
The reason for using `Map<String, String>` instead of `Map<String, ConfigValue>` is to be able to record `null` config values in a `ConcurrentHashMap`